src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -571,21 +583,21 @@ private static byte[] UrlEncodeNonAscii(byte[] bytes, int offset, int count)
                 return value;
             }
 
-            string? schemeAndAuthority;
-            string? path;
-            string? queryAndFragment;
+            Range schemeAndAuthority;
+            Range path;
+            Range queryAndFragment;
 
             if (!UriUtil.TrySplitUriForPathEncode(value, out schemeAndAuthority, out path, out queryAndFragment))
             {
                 // If the value is not a valid url, we treat it as a relative url.
                 // We don't need to extract query string from the url since UrlPathEncode()
                 // does not encode query string.
-                schemeAndAuthority = null;
-                path = value;
-                queryAndFragment = null;
+                schemeAndAuthority = default(Range);
+                path = Range.All;
+                queryAndFragment = default(Range);
             }
 
-            return schemeAndAuthority + UrlPathEncodeImpl(path) + queryAndFragment;
+            return string.Concat(value.AsSpan(schemeAndAuthority), UrlPathEncodeImpl(value.AsSpan(path).ToString()), value.AsSpan(queryAndFragment));
```

Comment:
This could be combined into the whole `UrlPathEncodeImpl` to avoid the substring allocations